File: desktop-v2/src/services/memoryAssistant.ts
/**
 * MemoryAssistant — TypeScript port of
 * `desktop/Desktop/Sources/ProactiveAssistants/Assistants/MemoryExtraction/MemoryAssistant.swift`.
 *
 * Hooks into the proactive frame stream from `proactiveAssistant.ts`. For each
 * frame that clears the exclusion + interval checks, we build a dedup-aware
 * prompt (last 20 extractions) and ask Gemini for a structured JSON response.
 * Accepted memories land in local SQLite first, then sync to the backend.
 *
 * Simpler than TaskAssistant — no tool-calling loop. Matches Swift.
 */

import { invoke } from "@tauri-apps/api/core";
import { api } from "@/services/api";
import { notify } from "@/services/notifications";
import {
  CapturedFrame,
  addFrameListener,
  startMonitoring,
  stopMonitoring,
} from "@/services/proactiveAssistant";
import {
  isAppAllowed,
  useMemoryAssistantSettings,
} from "@/services/memoryAssistantSettings";
import { useMemoryStore } from "@/stores/memoryStore";

export type MemoryCategory = "system" | "interesting";

export interface ExtractedMemory {
  content: string;
  category: MemoryCategory;
  source_app: string;
  confidence: number;
}

export interface MemoryExtractionResult {
  hasNewMemory: boolean;
  memories: ExtractedMemory[];
  contextSummary: string;
  currentActivity: string;
}

interface GeminiPart {
  text?: string;
  inlineData?: { mimeType: string; data: string };
}

interface GeminiResponse {
  candidates?: Array<{
    content?: { parts?: GeminiPart[] };
    finishReason?: string;
  }>;
}

const GEMINI_MODEL = "gemini-pro-latest";
const GEMINI_PATH = `/v1/proxy/gemini/models/${GEMINI_MODEL}:generateContent`;

const MAX_PREVIOUS_MEMORIES = 20;
const previousMemories: ExtractedMemory[] = [];

let lastAnalysisTime = 0;
let inflight = false;
let unsubscribeFrame: (() => void) | null = null;

const RESPONSE_SCHEMA = {
  type: "object",
  properties: {
    has_new_memory: {
      type: "boolean",
      description: "True if new memories were found",
    },
    memories: {
      type: "array",
      description: "Array of extracted memories (0-1 max)",
      items: {
        type: "object",
        properties: {
          content: { type: "string", description: "The memory content (max 15 words)" },
          category: {
            type: "string",
            enum: ["system", "interesting"],
            description: "Memory category",
          },
          source_app: { type: "string", description: "App where memory was found" },
          confidence: { type: "number", description: "Confidence score 0.0-1.0" },
        },
        required: ["content", "category", "source_app", "confidence"],
      },
    },
    context_summary: {
      type: "string",
      description: "Brief summary of what user is looking at",
    },
    current_activity: {
      type: "string",
      description: "High-level description of user's activity",
    },
  },
  required: ["has_new_memory", "memories", "context_summary", "current_activity"],
};

function buildPrompt(appName: string): string {
  let prompt = `Analyze this screenshot from ${appName}.\n\n`;
  if (previousMemories.length > 0) {
    prompt +=
      "RECENTLY EXTRACTED MEMORIES (do not re-extract these or semantically similar ones):\n";
    previousMemories.forEach((m, i) => {
      prompt += `${i + 1}. [${m.category}] ${m.content}\n`;
    });
    prompt += "\nLook for NEW memories that are NOT already in the list above.";
  } else {
    prompt +=
      "Look for memories to extract (system facts about the user, or interesting wisdom from others).";
  }
  return prompt;
}

async function callGemini(
  systemPrompt: string,
  prompt: string,
  imageBase64: string,
): Promise<string | null> {
  const body = {
    system_instruction: { parts: [{ text: systemPrompt }] },
    contents: [
      {
        role: "user",
        parts: [
          { text: prompt },
          { inlineData: { mimeType: "image/jpeg", data: imageBase64 } },
        ],
      },
    ],
    generationConfig: {
      maxOutputTokens: 2048,
      temperature: 0.2,
      response_mime_type: "application/json",
      response_schema: RESPONSE_SCHEMA,
    },
  };
  const resp = await api.post<GeminiResponse>(GEMINI_PATH, body);
  const text = resp?.candidates?.[0]?.content?.parts?.[0]?.text;
  return text ?? null;
}

function parseResult(raw: string): MemoryExtractionResult | null {
  try {
    const parsed = JSON.parse(raw) as {
      has_new_memory?: boolean;
      memories?: Array<Partial<ExtractedMemory>>;
      context_summary?: string;
      current_activity?: string;
    };
    const memories: ExtractedMemory[] = Array.isArray(parsed.memories)
      ? parsed.memories
          .filter(
            (m): m is ExtractedMemory =>
              typeof m?.content === "string" &&
              (m.category === "system" || m.category === "interesting") &&
              typeof m.source_app === "string" &&
              typeof m.confidence === "number",
          )
          .map((m) => ({
            content: m.content,
            category: m.category,
            source_app: m.source_app,
            confidence: m.confidence,
          }))
      : [];
    return {
      hasNewMemory: Boolean(parsed.has_new_memory),
      memories,
      contextSummary: String(parsed.context_summary ?? ""),
      currentActivity: String(parsed.current_activity ?? ""),
    };
  } catch (err) {
    console.warn("[MemoryAssistant] malformed JSON from Gemini:", err);
    return null;
  }
}

export async function extractMemories(
  frame: CapturedFrame,
): Promise<MemoryExtractionResult | null> {
  const settings = useMemoryAssistantSettings.getState();
  const prompt = buildPrompt(frame.appName);

  let raw: string | null;
  try {
    raw = await callGemini(settings.analysisPrompt, prompt, frame.imageBase64);
  } catch (err) {
    console.warn("[MemoryAssistant] Gemini call failed:", err);
    return null;
  }
  if (!raw) return null;
  return parseResult(raw);
}

interface BackendMemoryResponse {
  id: string;
  message?: string;
}

async function persistMemory(
  memory: ExtractedMemory,
  frame: CapturedFrame,
  result: MemoryExtractionResult,
): Promise<void> {
  const id =
    typeof crypto !== "undefined" && typeof crypto.randomUUID === "function"
      ? crypto.randomUUID()
      : `mem_${Date.now()}_${Math.floor(Math.random() * 1e9)}`;

  try {
    await invoke<string>("insert_memory", {
      input: {
        id,
        content: memory.content,
        category: memory.category,
        visibility: "private",
        confidence: memory.confidence,
        source_app: memory.source_app,
        window_title: frame.windowTitle,
        context_summary: result.contextSummary,
        current_activity: result.currentActivity,
        headline: null,
        reasoning: null,
        tags_json: null,
        screenshot_id: frame.dbId,
      },
    });
  } catch (err) {
    console.warn("[MemoryAssistant] local insert failed:", err);
    return;
  }

  const payload = {
    content: memory.content,
    visibility: "private",
    category: memory.category,
    confidence: memory.confidence,
    source_app: memory.source_app,
    context_summary: result.contextSummary,
    tags: [] as string[],
    reasoning: "",
    current_activity: result.currentActivity,
    source: "desktop",
    window_title: frame.windowTitle,
    headline: "",
  };

  try {
    const resp = await api.post<BackendMemoryResponse>("/v3/memories", payload);
    if (resp?.id) {
      try {
        await invoke("set_memory_backend_id", { id, backendId: resp.id });
      } catch (err) {
        console.warn("[MemoryAssistant] set_memory_backend_id failed:", err);
      }
    }
  } catch (err) {
    console.warn("[MemoryAssistant] backend sync failed:", err);
  }

  try {
    await useMemoryStore.getState().loadMemories();
  } catch (err) {
    console.warn("[MemoryAssistant] loadMemories refresh failed:", err);
  }
}

async function handleFrame(frame: CapturedFrame): Promise<void> {
  if (inflight) return;
  const settings = useMemoryAssistantSettings.getState();
  if (!settings.enabled) return;
  if (!isAppAllowed(frame.appName)) {
    return;
  }

  const intervalMs = settings.extractionIntervalSeconds * 1000;
  if (Date.now() - lastAnalysisTime < intervalMs) return;

  inflight = true;
  lastAnalysisTime = Date.now();
  try {
    console.info(
      `[MemoryAssistant] analyzing frame app=${frame.appName}`,
    );
    const result = await extractMemories(frame);
    if (!result || !result.hasNewMemory || result.memories.length === 0) {
      return;
    }

    const memory = result.memories[0];
    if (!memory) return;

    if (memory.confidence < settings.minConfidence) {
      console.info(
        `[MemoryAssistant] filtered ${(memory.confidence * 100).toFixed(0)}% < ${(settings.minConfidence * 100).toFixed(0)}%: "${memory.content}"`,
      );
      return;
    }

    console.info(
      `[MemoryAssistant] extracted [${memory.category}] "${memory.content}" conf=${memory.confidence.toFixed(2)}`,
    );

    previousMemories.unshift(memory);
    if (previousMemories.length > MAX_PREVIOUS_MEMORIES) {
      previousMemories.pop();
    }

    await persistMemory(memory, frame, result);

    if (settings.notificationsEnabled) {
      const title = memory.category === "interesting" ? "Wisdom captured" : "Memory saved";
      void notify(title, memory.content);
    }
  } catch (err) {
    console.warn("[MemoryAssistant] handleFrame error:", err);
  } finally {
    inflight = false;
  }
}

/** Start listening for proactive frames. Idempotent. */
export function initMemoryAssistant(): void {
  if (unsubscribeFrame) return;
  const off = addFrameListener((frame) => {
    void handleFrame(frame);
  });
  startMonitoring();
  unsubscribeFrame = () => {
    off();
    stopMonitoring();
  };
  console.info("[MemoryAssistant] started");
}

/** Stop listening and reset state. */
export function stopMemoryAssistant(): void {
  if (unsubscribeFrame) {
    unsubscribeFrame();
    unsubscribeFrame = null;
  }
  lastAnalysisTime = 0;
  inflight = false;
  previousMemories.length = 0;
  console.info("[MemoryAssistant] stopped");
}

File: desktop/macos/Desktop/Sources/ProactiveAssistants/Assistants/MemoryExtraction/MemoryAssistant.swift
import Foundation

/// Boxes a non-Sendable `[String: Any]` event payload so it can be sent into a
/// `@MainActor` task. The dictionary is built once and only read on the main
/// actor, so the unchecked Sendable conformance is sound.
private struct EventPayloadBox: @unchecked Sendable {
  let value: [String: Any]
}

/// Memory extraction assistant that identifies facts and wisdom from screen content
actor MemoryAssistant: ProactiveAssistant {
  typealias ExtractionOverride =
    @Sendable (_ jpegData: Data, _ appName: String) async throws -> MemoryExtractionResult?

  // MARK: - ProactiveAssistant Protocol

  nonisolated let identifier = "memory-extraction"
  nonisolated let displayName = "Memory Extractor"

  var isEnabled: Bool {
    get async {
      await MainActor.run {
        // Gate the Gemini screen analysis on notifications (off by default) — no
        // notification, no Gemini call. Re-enabling notifications resumes analysis.
        MemoryAssistantSettings.shared.isEnabled
          && MemoryAssistantSettings.shared.notificationsEnabled
      }
    }
  }

  // MARK: - Properties

  private let geminiClient: GeminiClient?
  private let extractionOverride: ExtractionOverride?
  private let durabilityPipeline: MemoryAssistantDurabilityPipeline
  private var isRunning = false
  private var lastAnalysisTime: Date = .distantPast
  private var previousMemories: [ExtractedMemory] = []  // Last 20 extracted memories for deduplication
  private let maxPreviousMemories = 20
  private var currentApp: String?
  private var pendingFrame: CapturedFrame?
  private var processingTask: Task<Void, Never>?
  private let frameSignal: AsyncStream<Void>
  private let frameSignalContinuation: AsyncStream<Void>.Continuation

  /// Get the current system prompt from settings (accessed on MainActor for thread safety)
  private var systemPrompt: String {
    get async {
      await MainActor.run {
        MemoryAssistantSettings.shared.analysisPrompt
      }
    }
  }

  /// Get the extraction interval from settings
  private var extractionInterval: TimeInterval {
    get async {
      await MainActor.run {
        MemoryAssistantSettings.shared.extractionInterval
      }
    }
  }

  /// Get the minimum confidence threshold from settings
  private var minConfidence: Double {
    get async {
      await MainActor.run {
        MemoryAssistantSettings.shared.minConfidence
      }
    }
  }

  // MARK: - Initialization

  init(apiKey: String? = nil) throws {
    // Flash-Lite: vision-capable, off the Vertex PT reservation, and measurably more
    // prompt-compliant than Flash on this lane (see ModelQoS.Gemini.lightweight).
    self.geminiClient = try GeminiClient(
      apiKey: apiKey, model: ModelQoS.Gemini.lightweight, workload: .extraction)
    self.extractionOverride = nil
    self.durabilityPipeline = MemoryAssistantDurabilityPipeline(
      runner: MemoryAssistantProductionDurability(operations: MemoryAssistantLiveDurabilityOperations())
    )

    let (stream, continuation) = AsyncStream.makeStream(of: Void.self, bufferingPolicy: .bufferingNewest(1))
    self.frameSignal = stream
    self.frameSignalContinuation = continuation

    // Start processing loop
    Task {
      await self.startProcessing()
    }
  }

  /// Hermetic analysis-attempt initializer. It uses the same `processFrame`
  /// catch and analytics boundary as production without constructing a live
  /// Gemini client or starting the background frame loop.
  init(
    extractionOverride: @escaping ExtractionOverride,
    durabilityRunner: any MemoryAssistantDurabilityRunning
  ) {
    self.geminiClient = nil
    self.extractionOverride = extractionOverride
    self.durabilityPipeline = MemoryAssistantDurabilityPipeline(runner: durabilityRunner)

    let (stream, continuation) = AsyncStream.makeStream(of: Void.self, bufferingPolicy: .bufferingNewest(1))
    self.frameSignal = stream
    self.frameSignalContinuation = continuation
  }

  // MARK: - Processing

  private func startProcessing() {
    isRunning = true
    processingTask = Task {
      await processLoop()
    }
  }

  private func processLoop() async {
    log("Memory assistant started")

    for await _ in frameSignal {
      guard isRunning else { break }
      guard pendingFrame != nil else { continue }

      // Wait until the extraction interval has passed
      let interval = await extractionInterval
      let timeSinceLastAnalysis = Date().timeIntervalSince(lastAnalysisTime)
      if timeSinceLastAnalysis < interval {
        let remaining = interval - timeSinceLastAnalysis
        try? await Task.sleep(nanoseconds: UInt64(remaining * 1_000_000_000))
      }

      // Grab the latest frame (may have been updated or cleared during sleep)
      guard let frame = pendingFrame else { continue }
      let waited = Date().timeIntervalSince(lastAnalysisTime)
      log("Memory: Starting analysis (interval: \(Int(interval))s, waited: \(Int(waited))s)")
      pendingFrame = nil
      lastAnalysisTime = Date()
      await processFrame(frame)
    }

    log("Memory assistant stopped")
  }

  // MARK: - ProactiveAssistant Protocol Methods

  func shouldAnalyze(frameNumber: Int, timeSinceLastAnalysis: TimeInterval) -> Bool {
    // Memory assistant analyzes less frequently - every N seconds
    // The actual interval is checked in the processing loop
    // Here we just accept frames to store the latest one
    return true
  }

  func analyze(frame: CapturedFrame) async -> AssistantResult? {
    // Skip apps excluded from memory extraction (built-in + user's custom list)
    let exclusion = await MainActor.run {
      let settings = MemoryAssistantSettings.shared
      return (
        settings.isAppExcluded(frame.appName),
        settings.excludedHosts,
        NegativeFeedbackRemediationFeature.isEnabled
      )
    }
    if exclusion.0 {
      log("Memory: Skipping excluded app '\(frame.appName)'")
      return nil
    }
    if exclusion.2 {
      let snapshot = WorkHistoryHandleExtractor.liveSnapshot(
        appName: frame.appName, windowTitle: frame.windowTitle)
      let urlString = snapshot.browserURL?.absoluteString ?? snapshot.documentURL?.absoluteString
      if MemoryHostExclusion.isExcluded(urlString: urlString, excludedHosts: exclusion.1) {
        log("Memory: Skipping excluded host '\(urlString ?? "")'")
        return nil
      }
    }

    // Store the latest frame - we'll process it when the interval has passed
    let hadPending = pendingFrame != nil
    pendingFrame = frame
    if !hadPending {
      log("Memory: Received frame from \(frame.appName), queued for analysis")
    }
    // Signal the processing loop that a frame is available
    frameSignalContinuation.yield()
    return nil
  }

  func handleResult(_ result: AssistantResult, sendEvent: @escaping @Sendable (String, [String: Any]) -> Void) async {
    // This method is required by protocol but we use handleResultWithScreenshot instead
    guard let memoryResult = result as? MemoryExtractionResult else { return }
    guard let ownerID = RuntimeOwnerIdentity.currentOwnerId() else { return }
    await handleResultWithScreenshot(
      memoryResult,
      ownerID: ownerID,
      screenshotId: nil,
      sendEvent: sendEvent
    )
  }

  /// Handle result with screenshot ID for SQLite storage
  private func handleResultWithScreenshot(
    _ memoryResult: MemoryExtractionResult,
    ownerID: String,
    screenshotId: Int64?,
    windowTitle: String? = nil,
    sendEvent: @escaping (String, [String: Any]) -> Void
  ) async {
    guard RuntimeOwnerIdentity.currentOwnerId() == ownerID else { return }
    // Check if AI found new memories
    guard memoryResult.hasNewMemory, !memoryResult.memories.isEmpty else {
      await recordAnalysisOutcome(.noNewMemory, ownerID: ownerID)
      return
    }

    // Get min confidence threshold
    let threshold = await minConfidence
    guard RuntimeOwnerIdentity.currentOwnerId() == ownerID else { return }

    // Only process the first memory (max 1 per analysis)
    guard let memory = memoryResult.memories.first else { return }

    let confidencePercent = Int(memory.confidence * 100)

    // Check confidence threshold
    guard memory.confidence >= threshold else {
      log("Memory: [\(confidencePercent)% < \(Int(threshold * 100))%] Filtered: \"\(memory.content)\"")
      await recordAnalysisOutcome(.filteredLowConfidence, confidence: memory.confidence, ownerID: ownerID)
      return
    }

    let subjectGateEnabled = await MainActor.run { NegativeFeedbackRemediationFeature.isEnabled }
    if subjectGateEnabled, !MemoryAdmissionGate.admits(memory) {
      log("Memory: subject-admission refused: \"\(memory.content)\"")
      await recordAnalysisOutcome(.filteredSubjectAdmission, confidence: memory.confidence, ownerID: ownerID)
      return
    }

    log("Memory: [\(confidencePercent)% conf.] [\(memory.category.rawValue)] \"\(memory.content)\"")

    // Add to previous memories (keep last 20 for deduplication context)
    previousMemories.insert(memory, at: 0)
    if previousMemories.count > maxPreviousMemories {
      previousMemories.removeLast()
    }

    let durability = await durabilityPipeline.persistSyncAndEmit(
      MemoryAssistantDurabilityRequest(
        memory: memory,
        screenshotId: screenshotId,
        contextSummary: memoryResult.contextSummary,
        windowTitle: windowTitle,
        ownerID: ownerID
      ),
      confidence: memory.confidence
    )
    guard RuntimeOwnerIdentity.currentOwnerId() == ownerID else { return }
    // A failed insert never reaches backend sync and must never be reported as
    // the historical extraction success. The pipeline already recorded exactly
    // one closed analysis terminal (and historical success where appropriate).
    guard durability.shouldEmitMemoryExtracted else { return }
    guard RuntimeOwnerIdentity.currentOwnerId() == ownerID else { return }

    // Send notification if enabled
    let notificationsEnabled = await MainActor.run {
      MemoryAssistantSettings.shared.notificationsEnabled
    }
    guard RuntimeOwnerIdentity.currentOwnerId() == ownerID else { return }
    if notificationsEnabled {
      await sendMemoryNotification(
        ownerID: ownerID,
        memory: memory,
        result: memoryResult,
        windowTitle: windowTitle
      )
    }
    guard RuntimeOwnerIdentity.currentOwnerId() == ownerID else { return }

    // Send event to Flutter
    sendEvent(
      "memoryExtracted",
      [
        "assistant": identifier,
        "memory": memory.toDictionary(),
        "contextSummary": memoryResult.contextSummary,
      ])
  }

  /// Record one analysis-outcome telemetry event, gated on the owner not having
  /// switched mid-analysis (a prior owner's attempt must never be attributed to the
  /// current PostHog identity). Supplements — never replaces — `Memory Extracted`,
  /// which continues to fire only on the local-persistence success terminal below.
  private func recordAnalysisOutcome(
    _ outcome: MemoryAssistantTelemetry.AnalysisOutcome,
    confidence: Double? = nil,
    ownerID: String
  ) async {
    await MainActor.run {
      guard RuntimeOwnerIdentity.currentOwnerId() == ownerID else { return }
      AnalyticsManager.shared.memoryAssistantAnalysisRun(outcome: outcome, confidence: confidence)
    }
  }

  /// Send a notification for the extracted memory
  private func sendMemoryNotification(
    ownerID: String,
    memory: ExtractedMemory,
    result: MemoryExtractionResult,
    windowTitle: String?
  ) async {
    // One category, one name: every memory notification presents as "Memory" — the
    // "Wisdom Captured" variant read as an unclassifiable notification type.
    // The category lives on the badge / system-banner title; the body is the
    // memory itself. Prefixing "New memory:" stacked a third copy of the same word.
    let title = "Memory Saved"
    let message = memory.content
    let context = FloatingBarNotificationContext(
      sourceTitle: title,
      assistantId: identifier,
      sourceApp: memory.sourceApp.isEmpty ? nil : memory.sourceApp,
      windowTitle: windowTitle,
      contextSummary: result.contextSummary,
      currentActivity: result.currentActivity,
      reasoning: nil,
      detail: memory.content
    )

    await MainActor.run {
      NotificationService.shared.sendNotification(
        ownerID: ownerID,
        title: title,
        message: message,
        assistantId: identifier,
        context: context
      )
    }
  }

  func onAppSwitch(newApp: String) async {
    if newApp != currentApp {
      if let currentApp = currentApp {
        log("Memory: APP SWITCH: \(currentApp) -> \(newApp)")
      } else {
        log("Memory: Active app: \(newApp)")
      }
      currentApp = newApp
      // Don't clear previous memories on app switch - we want to track across apps
    }
  }

  func clearPendingWork() async {
    pendingFrame = nil
    log("Memory: Cleared pending frame")
  }

  func stop() async {
    isRunning = false
    frameSignalContinuation.finish()
    processingTask?.cancel()
    pendingFrame = nil
  }

  // MARK: - Analysis

  func processFrame(_ frame: CapturedFrame) async {
    guard let ownerID = RuntimeOwnerIdentity.currentOwnerId() else { return }
    let enabled = await isEnabled
    guard enabled else {
      log("Memory: Skipping analysis (disabled)")
      return
    }

    log("Memory: Analyzing frame from \(frame.appName)...")
    do {
      guard let result = try await extractMemories(from: frame.jpegData, appName: frame.appName) else {
        log("Memory: Analysis returned no result")
        await recordAnalysisOutcome(.analysisFailed, ownerID: ownerID)
        return
      }

      log(
        "Memory: Analysis complete - hasNewMemory: \(result.hasNewMemory), count: \(result.memories.count), context: \(result.contextSummary)"
      )

      // Handle the result with screenshot ID for SQLite storage
      await handleResultWithScreenshot(
        result,
        ownerID: ownerID,
        screenshotId: frame.screenshotId,
        windowTitle: frame.windowTitle
      ) { type, data in
        let payload = EventPayloadBox(value: data)
        Task { @MainActor in
          AssistantCoordinator.shared.sendEvent(type: type, data: payload.value)
        }
      }
    } catch {
      logError("Memory extraction error", error: error)
      await recordAnalysisOutcome(.analysisFailed, ownerID: ownerID)
    }
  }

  private func extractMemories(from jpegData: Data, appName: String) async throws -> MemoryExtractionResult? {
    if let extractionOverride {
      return try await extractionOverride(jpegData, appName)
    }
    guard let geminiClient else {
      return nil
    }

    // Build context with previous memories for deduplication
    var prompt = "Analyze this screenshot from \(appName).\n\n"

    if !previousMemories.isEmpty {
      prompt += "RECENTLY EXTRACTED MEMORIES (do not re-extract these or semantically similar ones):\n"
      for (index, memory) in previousMemories.enumerated() {
        prompt += "\(index + 1). [\(memory.category.rawValue)] \(memory.content)\n"
      }
      prompt += "\nLook for NEW memories that are NOT already in the list above."
    } else {
      prompt += "Look for memories to extract (system facts about the user)."
    }

    // Get current system prompt from settings
    let currentSystemPrompt = await systemPrompt

    // Build response schema for memory extraction
    let subjectGateEnabled = await MainActor.run { NegativeFeedbackRemediationFeature.isEnabled }
    var memoryProperties: [String: GeminiRequest.GenerationConfig.ResponseSchema.Property] = [
      "content": .init(type: "string", description: "The memory content (max 15 words)"),
      "category": .init(
        type: "string",
        enum: subjectGateEnabled ? ["system"] : ["system", "interesting"],
        description: "Memory category"),
      "source_app": .init(type: "string", description: "App where memory was found"),
      "confidence": .init(type: "number", description: "Confidence score 0.0-1.0"),
    ]
    var requiredFields = ["content", "category", "source_app", "confidence"]
    if subjectGateEnabled {
      memoryProperties["subject_scope"] = .init(
        type: "string",
        enum: ["primary_user", "third_party", "artifact"],
        description: "Who the memory is about")
      memoryProperties["subject_evidence"] = .init(
        type: "string",
        enum: ["user_authored", "addressed_to_user", "rendered_content", "ui_chrome"],
        description: "How the subject was established")
      memoryProperties["contains_credential_or_identifier"] = .init(
        type: "boolean",
        description: "True if the memory contains a password, API key, or account identifier")
      requiredFields.append(contentsOf: [
        "subject_scope", "subject_evidence", "contains_credential_or_identifier",
      ])
    }

    let responseSchema = GeminiRequest.GenerationConfig.ResponseSchema(
      type: "object",
      properties: [
        "has_new_memory": .init(type: "boolean", description: "True if new memories were found"),
        "memories": .init(
          type: "array",
          description: "Array of extracted memories (0-3 max)",
          items: .init(
            type: "object",
            properties: memoryProperties,
            required: requiredFields
          )
        ),
        "context_summary": .init(type: "string", description: "Brief summary of what user is looking at"),
        "current_activity": .init(type: "string", description: "High-level description of user's activity"),
      ],
      required: ["has_new_memory", "memories", "context_summary", "current_activity"]
    )

    let responseText = try await geminiClient.sendRequest(
      prompt: prompt,
      imageData: jpegData,
      systemPrompt: currentSystemPrompt,
      responseSchema: responseSchema
    )

    return try JSONDecoder().decode(MemoryExtractionResult.self, from: Data(responseText.utf8))
  }
}
